Support for (optional) ack timeout and nack delay for consumers

dionjansen opened the pull request:
Implements: #45 , #46 

## General comments

* I've added a reference to Microsoft.Bcl.AsyncInterfaces to support the IDisposable interface in net5.0. I only had this issue in VSCode for the test and sample projects. This does not influence the tests or running samples on VSC (on mac).
* This is still a draft mostly to get the discussion going. 
* I would like to add some more unit tests though this is a bit hard without a proper mocking framework, would you be open to introducing something like Moq to allow testing classes in a shallow way? Or do you rather have I follow a different unit test strategy?

## Discussion
I think we can split up the discussion in 2 parts.

1. The integration in the rest of the lib: where what is created, started and what is optional, general structure/ logic of the lib
2. The implementation of the tracker: performance concerns, disposing

@blankensteiner I think firstly I would like to get some thoughts on the current implementation, I mostly focussed now on setting things up so I don't break existing processes:

### Integration
1. Right now I create a `IMessageAcksTracker` instance in the `PulsarClient`, though when clients do not have the consumer configured to use the tracking, a dummy `InactiveMessageAcksTracker` is passed. This instance is passed to the `ConsumerChannelFactory` and used when a channel is created. Does this setup makes sense to you?
2. The `ConsumerChannel` now expects a `MessageQueue` which wraps the tracker and the `AsyncQueue`. The channel now informs the tracker through this queue when acking (and later nacking which I can add later) and the dequeue method automatically starts tracking messages received.
2. When the tracker is started it runs indefinitely I find it hard to find a good place to start this thread, also since I need the consumer to call `RedeliverUnacknowledgedMessages`. Any ideas on how to improve this pattern? Perhaps a static method like some of the `StateMonitor` methods.
3. I wonder how batched messages fit into all this, does a batch have a single message id? And should I be able to just redeliver that message id to release the batch back to the broker?

### Implementation
1. The `MessageAcksTracker` uses a polling mechanism to re-check for timed out messages (either due to being unacked for too long or the nack delay has been exceeded). Is this (generally) what you were thinking about too? Alternatively I could think of an approach where polling is done through a [Timer](https://docs.microsoft.com/en-us/dotnet/api/system.threading.timer?view=net-5.0). Or we could create individual Tasks for each added message to the tracker but I'm concerned of the overhead created by this.
2. Atm I haven't considered the scenario yet that only a nack delay is configured and not an ack timeout in which case we will not have to track all dequeued messages.

blankensteiner commented:
Hi @dionjansen 
Thanks for the PR!
I'll try and answer the best I can :-)

> I've added a reference to Microsoft.Bcl.AsyncInterfaces to support the IDisposable interface in net5.0. I only had this issue in VSCode for the test and sample projects. This does not influence the tests or running samples on VSC (on mac).

We have tried adding support for Visual Studio Code before, but sadly it's just not a pleasant road to go down. I have no idea why Microsoft has created two IDE's for C# and doesn't ensure that they behave the same.
Visual Studio Code will create nonsense warnings and require unnecessary changes to the code-base and therefore it's not supported for developing DotPulsar. Visual Studio Community Edition and the commercial offerings and Rider are supported, so you have to use one of those.

> I would like to add some more unit tests though this is a bit hard without a proper mocking framework, would you be open to introducing something like Moq to allow testing classes in a shallow way? Or do you rather have I follow a different unit test strategy?

Feel free to add one or more of these (if you need them):

- AutoFixture
- AutoFixture.AutoNSubstitute
- AutoFixture.Xunit2
- NSubstitute

> Right now I create a IMessageAcksTracker instance in the PulsarClient, though when clients do not have the consumer configured to use the tracking, a dummy InactiveMessageAcksTracker is passed. This instance is passed to the ConsumerChannelFactory and used when a channel is created. Does this setup makes sense to you?

Yes.

> When the tracker is started it runs indefinitely I find it hard to find a good place to start this thread, also since I need the consumer to call RedeliverUnacknowledgedMessages. Any ideas on how to improve this pattern? Perhaps a static method like some of the StateMonitor methods.

The question is if you actually need to consumer or just the consumer channel. The tracking should start and end together with the MessageQueue/ConsumerChannel.

> I wonder how batched messages fit into all this, does a batch have a single message id? And should I be able to just redeliver that message id to release the batch back to the broker?

That's a really good question. Bookkeeper stores batched messages as one, so if you have a batched message consisting of 5 messages and you ack 4 of them, but the last times out and you ask the broker to redeliver, you will get the entire batch again. You could keep track of this, but it will hurt performance. I don't know what the other clients do here, but maybe you could test that?

> The MessageAcksTracker uses a polling mechanism to re-check for timed out messages (either due to being unacked for too long or the nack delay has been exceeded). Is this (generally) what you were thinking about too? Alternatively I could think of an approach where polling is done through a Timer. Or we could create individual Tasks for each added message to the tracker but I'm concerned of the overhead created by this.

I agree that a timer/task per message will hurt performance too much. Have one task for the entire Consumer/MessageQueue is the right solution. Waking up and looking at what needs to be redelivered. Here we need to find a thread-safe and performant way of storing and accessing this information.

> Atm I haven't considered the scenario yet that only a nack delay is configured and not an ack timeout in which case we will not have to track all dequeued messages.

Just a boolean check on dequeue and ack? We also need to handle cumulative acknowledgment.

Consider this ackTimeout implementation. First ackTimeout should be giving as a TimeSpan (instead of a an int or long as milli- or micro-seconds).

When a message is dequeued, and if we have an ackTimeout, then we store the MessageId and StopWatch.GetTimestamp() in an "AwaitingAck" struct in a ConcurrencyQueue, let's call it "AwaitingAcks". Other suggestions for concurrent collections with fast insertion are welcome.

When a message is acknowledged, and if we have an ackTimeout, then we store the acks instead of removing them from "AwaitingAcks". If we want to remove them right away, then we need the "AwaitingAcks" collection to support both iteration and random deletion. We only need to store the highest cumulative ack we see (if there is one) and the MessageIds not included by that cumulative ack.

When the ackTracker wakes up and has calculated what ackTimeout is in StopWatch ticks (those are not the same as TimeSpan ticks). It will call StopWatch.GetTimestamp(). We will now TryPeek and Dequeue from "AwaitingAck" for as long as the tracker timestamp -  AwaitingAck.Timestamp is larger than the calculated timeout.
If the MessageId is not acknowledged, it's added to a CommandRedeliverUnacknowledgedMessages (that we are reusing) and then send it (if MessageIds were added).
If the MessageId was acknowledged, then we can remove that MessageId from the AckedMessageIds.

Consider this nackTimeout implementation.

When a message is nacked, we added the MessageId(s) to a  CommandRedeliverUnacknowledgedMessages that we are reusing (should our "RedeliverUnacknowledgedMessages" taking an enumerable of messageId actually have been called "NegativeAcknowledge"?).

When the nackTracker wakes up it will check if the CommandRedeliverUnacknowledgedMessages has MessageIds and if yes, then send it.

Writing such a detailed implementation description was not what I intended, but when I first get going.... :-D
Anyway, if it is unclear, then I can try and make those classes/structs and push them to master.

dionjansen commented:
@blankensteiner I've tried to follow your suggestions and simplify the implementation a bit: I focussed just on the unacked message tracker. Perhaps it is anyway a good idea to separate these two mechanisms (nack tracking and unacked tracking) since they are configured independently of each other.

> Feel free to add one or more of these (if you need them):
> 
> * AutoFixture
> * AutoFixture.AutoNSubstitute
> * AutoFixture.Xunit2
> * NSubstitute
> 

Added `NSubstitute` and `AutoFixture.AutoNSubstitute`. I started testing the unacked tracker with this but I'm not sure I'm using the correct pattern to verify if messages are being redelivered under different test conditions. Since I don't see any other unit tests that tests internal classes in this way I'm not sure if this strategy agrees with the rest of the lib, so let me know what you think.

> The question is if you actually need to consumer or just the consumer channel. The tracking should start and end together with the MessageQueue/ConsumerChannel.

I am starting the thread now in the Pulsar client https://github.com/apache/pulsar-dotpulsar/blob/26cd957a56b6349f9e37bf76bd37f92a7a7e0970/src/DotPulsar/PulsarClient.cs#L81-L84

The tracker is then passed to the channel factory, so it can be passed to a message queue that is passed to the channel when created. The tracker loop is stopped when disposed which occurs when the message queue is disposed, which in terms occurs when the channel is disposed. I'm still using `IConsumer` for start in order to avoid duplicate implementation of `RedeliverUnacknowledgedMessages` in the consumer.

> That's a really good question. Bookkeeper stores batched messages as one, so if you have a batched message consisting of 5 messages and you ack 4 of them, but the last times out and you ask the broker to redeliver, you will get the entire batch again. You could keep track of this, but it will hurt performance. I don't know what the other clients do here, but maybe you could test that?

From what I can see in the [`ConsumerImpl` in java](https://github.com/apache/pulsar/blob/6926180966f45eb9c1499b7f0eb32ea2a1368fd6/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L1437-L1448) in case of batch messages only one item is put in the tracker for a batch message using `(ledgerId, entryId, partitionIndex)`. Then [when acking](https://github.com/apache/pulsar/blob/cc64889abe94d47f048e2f8e8fb10d6c37e695ec/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L602-L616) it looks like the batch is again treated as a single message, but only if `markAckForBatchMessage` returns false which indicates not all messages in the batch have been acked yet. I don't really see from this implementation how  they *"keep track"* of what to ack within the batch in this way, but there is a lot going one here that I can't make sense of. Does the adding removing mechanism I point out here to the tracker make any sense to you?

> Consider this ackTimeout implementation. First ackTimeout should be giving as a TimeSpan (instead of a an int or long as milli- or micro-seconds).

done, I kept the configuration options of the consumer to milliseconds though.

> When a message is dequeued, and if we have an ackTimeout, then we store the MessageId and StopWatch.GetTimestamp() in an "AwaitingAck" struct in a ConcurrencyQueue, let's call it "AwaitingAcks". Other suggestions for concurrent collections with fast insertion are welcome.

I was struggling a bit to create a comparable TimeSpan (since as you point out stopwatch ticks != timespan ticks). I followed [this](http://geekswithblogs.net/BlackRabbitCoder/archive/2012/01/12/c.net-little-pitfalls-stopwatch-ticks-are-not-timespan-ticks.aspx) article and concluded: https://github.com/apache/pulsar-dotpulsar/blob/f4725f5d81e8715a55ab3e4ea6791f903e9e9ad4/src/DotPulsar/Internal/UnackedMessageTracker.cs#L24-L25
Not casting frequency explicitly to double results in considerable loss of accuracy.

> When a message is acknowledged, and if we have an ackTimeout, then we store the acks instead of removing them from "AwaitingAcks". If we want to remove them right away, then we need the "AwaitingAcks" collection to support both iteration and random deletion. We only need to store the highest cumulative ack we see (if there is one) and the MessageIds not included by that cumulative ack.

How can I determine if there is a highest cumulative ack from a `MessageId` instance? And if we do this wouldn't we also need some kind of removal from the unacked list that removes until this highest value, like: [removeMessagesTill](https://github.com/apache/pulsar/blob/6926180966f45eb9c1499b7f0eb32ea2a1368fd6/pulsar-client/src/main/java/org/apache/pulsar/client/impl/UnAckedMessageTracker.java#L225)?

> 
> When the ackTracker wakes up and has calculated what ackTimeout is in StopWatch ticks (those are not the same as TimeSpan ticks). It will call StopWatch.GetTimestamp(). We will now TryPeek and Dequeue from "AwaitingAck" for as long as the tracker timestamp - AwaitingAck.Timestamp is larger than the calculated timeout.
> If the MessageId is not acknowledged, it's added to a CommandRedeliverUnacknowledgedMessages (that we are reusing) and then send it (if MessageIds were added).
> If the MessageId was acknowledged, then we can remove that MessageId from the AckedMessageIds.
> 

I tried to capture this in https://github.com/apache/pulsar-dotpulsar/blob/51f4a984a5819aeb91ebcd25345b495c87cc02a7/src/DotPulsar/Internal/UnackedMessageTracker.cs#L81-L91. The only thing I'm still unsure about is the accumulated acking.

> Consider this nackTimeout implementation.
> 
> When a message is nacked, we added the MessageId(s) to a CommandRedeliverUnacknowledgedMessages that we are reusing (should our "RedeliverUnacknowledgedMessages" taking an enumerable of messageId actually have been called "NegativeAcknowledge"?).
> 
> When the nackTracker wakes up it will check if the CommandRedeliverUnacknowledgedMessages has MessageIds and if yes, then send it.

I will implement the nack tracker if you are happy with the unacked tracker as it stands (which might also be refactored into one single tracker if two trackers are a performance concern to you).

> 
> Writing such a detailed implementation description was not what I intended, but when I first get going.... :-D
> Anyway, if it is unclear, then I can try and make those classes/structs and push them to master.

Let me know what you think, this is still a bit of a learning process for me both on the internals of this lib as well as C# / Pulsar details. Thanks in advance 👍 .

blankensteiner commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,172 @@
+        }
+
+        [Fact]
+        public async void Test_Start_Message()
```
The naming convention for tests is Method/Feature_[Given]State_[Should]ExpectedResult.
E.g From AsyncLockTests:
Dispose_GivenLockIsDisposedWhileItIsTaken_ShouldNotCompleteBeforeItIsReleased()

dionjansen replied:
Done in b57668b

blankensteiner commented:
Hi @dionjansen 
Before doing a deep dive into the implementation, I have some comments on little things that can quickly be fixed :-)

dionjansen commented on src/DotPulsar/Internal/Abstractions/IMessageAcksTracker.cs:
```diff
@@ -0,0 +1,13 @@
+namespace DotPulsar.Internal.Abstractions
```
Done https://github.com/apache/pulsar-dotpulsar/pull/67/commits/7ad89a20a97bfc9ff1770ee1ae2abdda952870f6

dionjansen commented on src/DotPulsar/Internal/Abstractions/IMessageAcksTracker.cs:
```diff
@@ -0,0 +1,13 @@
+    using DotPulsar.Abstractions;
+    using System.Threading;
+    using System.Threading.Tasks;
+    public interface IMessageAcksTracker<T>
```
cleaned up in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/Abstractions/IMessageAcksTracker.cs:
```diff
@@ -0,0 +1,13 @@
+        T Nack(T message);
+        Task StartTracker(IConsumer consumer, CancellationToken cancellationToken);
+    }
+}
```
cleaned up in 0119ca9

dionjansen commented on src/DotPulsar/Internal/Abstractions/IMessageQueue.cs:
```diff
@@ -0,0 +1,10 @@
+namespace DotPulsar.Internal.Abstractions
```
Done https://github.com/apache/pulsar-dotpulsar/pull/67/commits/7ad89a20a97bfc9ff1770ee1ae2abdda952870f6

dionjansen commented on src/DotPulsar/Internal/InactiveMessageAcksTracker.cs:
```diff
@@ -0,0 +1,30 @@
+namespace DotPulsar.Internal
```
Done 7ad89a2

dionjansen commented on src/DotPulsar/Internal/Abstractions/IUnackedMessageTracker.cs:
```diff
@@ -0,0 +1,16 @@
+﻿namespace DotPulsar.Internal.Abstractions
```
Done 7ad89a2

dionjansen commented on src/DotPulsar/Internal/InactiveUnackedMessageTracker.cs:
```diff
@@ -0,0 +1,31 @@
+﻿namespace DotPulsar.Internal
```
Done 7ad89a2

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+namespace DotPulsar.Internal
```
Done 7ad89a2

dionjansen commented on src/DotPulsar/Internal/InactiveMessageAcksTracker.cs:
```diff
@@ -0,0 +1,30 @@
+        public MessageId Ack(MessageId message) => message;
+        public MessageId Nack(MessageId message) => message;
+    }
+}
```
I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+    using System.Threading.Tasks;
+    using System.Diagnostics;
+
+    internal class Tracker
```
Got it 👍 . I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+    internal class Tracker
+    {
+        private readonly Stopwatch _timer;
+        private int maxTimeoutMs;
```
I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+        public Tracker(int timeoutMs)
+        {
+            maxTimeoutMs = timeoutMs;
+            _timer = new Stopwatch();
```
I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    // TODO add mechnism to stop tracker when disposed
+    public sealed class MessageAcksTracker : IMessageAcksTracker<MessageId>
```
I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+        public MessageId Add(MessageId message)
+        {
+            if (!_trackers.ContainsKey(message))
+            {
```
I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+        private readonly int _unackedTimeoutMs;
+        private readonly int _nackTimeoutMs;
+        private readonly int _trackerDelayMs;
+        public MessageAcksTracker(int unackedTimeoutMs, int nackTimeoutMs, int trackerDelayMs)
```
I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+
+            return message;
+        }
+        public MessageId Ack(MessageId message)
```
I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/MessageAcksTracker.cs:
```diff
@@ -0,0 +1,103 @@
+            return message;
+        }
+    }
+}
```
I've cleaned up this implementation in https://github.com/apache/pulsar-dotpulsar/pull/67/commits/0119ca95e00be46184f48ab847d5353cf7e6a74f

dionjansen commented on src/DotPulsar/Internal/InactiveUnackedMessageTracker.cs:
```diff
@@ -0,0 +1,31 @@
+﻿namespace DotPulsar.Internal
+{
+    using System.Threading;
```
Do you run code cleanup through VSC? I've used *Remove and sort using* but this orders it in the way I wrote it now.

![image](https://user-images.githubusercontent.com/5121647/103767382-cb9a6180-5020-11eb-8539-fa9df66d888a.png)

But I can't find a code cleanup action in my IDE:

![image](https://user-images.githubusercontent.com/5121647/103767916-cab5ff80-5021-11eb-99a3-6fecdc340406.png)

Also no mention in the [docs](https://docs.microsoft.com/en-us/visualstudio/mac/editor-behavior?view=vsmac-2019) from what I can find, are you using a different editor? For now I'll just keep the order by hand

dionjansen replied:
Done https://github.com/apache/pulsar-dotpulsar/pull/67/commits/a7d5f02704d2ab598377009bf1115794c1a6db2b

blankensteiner replied:
Ah. In my VS (Community at home and Profession at work, both on Windows) there is a "Analyze" in the toolbar with a "Clean up" submenu, allowing you to run a Code Cleanup on the solution.
The IDE should respect the .editconfig and place usings alphabetically.

dionjansen replied:
Ok, I clearly don't have that in my version, strange I'll have a look to see how to configure my editor to respect .editconfig

dionjansen commented on src/DotPulsar/Internal/InactiveUnackedMessageTracker.cs:
```diff
@@ -0,0 +1,31 @@
+        public void Dispose() {
+            return;
+        }
+
```
Done

dionjansen commented on src/DotPulsar/Internal/MessageQueue.cs:
```diff
@@ -0,0 +1,45 @@
+namespace DotPulsar.Internal
```
Done 7ad89a2

dionjansen commented on src/DotPulsar/Internal/UnackedMessageTracker.cs:
```diff
@@ -0,0 +1,101 @@
+﻿namespace DotPulsar.Internal
```
Done 7ad89a2

dionjansen commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,172 @@
+namespace DotPulsar.Tests.Internal
```
Done 7ad89a2

dionjansen commented on src/DotPulsar/Internal/InactiveMessageAcksTracker.cs:
```diff
@@ -0,0 +1,30 @@
+
+    public sealed class InactiveMessageAcksTracker : IMessageAcksTracker<MessageId>
+    {
+        public InactiveMessageAcksTracker() { }
```
Done https://github.com/apache/pulsar-dotpulsar/pull/67/commits/a7d5f02704d2ab598377009bf1115794c1a6db2b

dionjansen commented on src/DotPulsar/Internal/MessageQueue.cs:
```diff
@@ -0,0 +1,45 @@
+            _queue.Dispose();
+            _tracker.Dispose();
+        }
+        
```
Done 2edfa35

dionjansen commented on src/DotPulsar/Internal/MessageQueue.cs:
```diff
@@ -0,0 +1,45 @@
+        }
+        
+    }
+}
```
Done 2edfa35

dionjansen commented on src/DotPulsar/Internal/MessageQueue.cs:
```diff
@@ -0,0 +1,45 @@
+            _tracker.Add(new MessageId(message.MessageId));
+            return message;
+        }
+        public void Acknowledge(MessageId obj) => _tracker.Ack(obj);
```
Done 2edfa35

dionjansen commented on src/DotPulsar/Internal/MessageQueue.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        private readonly AsyncQueue<MessagePackage> _queue;
+        private readonly IUnackedMessageTracker _tracker;
+        public MessageQueue(AsyncQueue<MessagePackage> queue, IUnackedMessageTracker tracker)
```
Done 2edfa35

dionjansen commented on src/DotPulsar/Internal/Abstractions/IMessageQueue.cs:
```diff
@@ -0,0 +1,10 @@
+
+    public interface IMessageQueue : IDequeue<MessagePackage>, IDisposable
+    {
+        void Acknowledge(MessageId obj);
```
Done 2edfa35

dionjansen commented on src/DotPulsar/Internal/Abstractions/IMessageQueue.cs:
```diff
@@ -0,0 +1,10 @@
+        void Acknowledge(MessageId obj);
+        void NegativeAcknowledge(MessageId obj);
+    }
+}
```
Done 2edfa35

dionjansen commented on tests/DotPulsar.StressTests/DotPulsar.StressTests.csproj:
```diff
@@ -7,6 +7,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.3" />
+    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="5.0.0" />
```
ah yes done here https://github.com/apache/pulsar-dotpulsar/pull/67/commits/98850bb8a5c0e6ade58714cf6c8da00527d258db

dionjansen commented on src/DotPulsar/Internal/UnackedMessageTracker.cs:
```diff
@@ -0,0 +1,101 @@
+              CancellationTokenSource.CreateLinkedTokenSource(
+                  _cancellationTokenSource.Token, cancellationToken).Token;
+
+            return Task.Run(async () => {
```
I'm not sure I understand. Did you mean ***doesn't** really make much sense*? Shouldn't I always create a task so that the while loop isn't blocking the caller thread?

blankensteiner replied:
Ah, yes, sorry, meant to say that it doesn't make much sense.
In this case, you have an async method and if people await it, it will block no matter if you run the task or not, since you are returning the task.

dionjansen replied:
Ok fixed in f02166b I followed the same pattern as in StateMonitor.

dionjansen commented on src/DotPulsar/Internal/UnackedMessageTracker.cs:
```diff
@@ -0,0 +1,101 @@
+        private readonly ConcurrentQueue<AwaitingAck> _awaitingAcks;
+        private readonly List<MessageId> _acked;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+
```
Done b78ac54

dionjansen commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,172 @@
+    public class UnackedMessageTrackerTests
+    {
+        [Fact]
+        public void Test_Instance()
```
Done, removed in b57668b

dionjansen commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,172 @@
+        {
+            //Arrange
+            var messageId = MessageId.Latest;
+            var sw = new Stopwatch();
```
Updated in b57668b

dionjansen commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,172 @@
+            var tracker = new UnackedMessageTracker(TimeSpan.FromSeconds(10), TimeSpan.FromSeconds(1));
+            tracker.Should().BeOfType<UnackedMessageTracker>();
+        }
+
```
Fixed in b57668b

dionjansen commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,172 @@
+        private bool IsEquivalentIEnumerable<T>(IEnumerable<T> a, IEnumerable<T> b) =>
+            a.Count() == b.Count() && a.Zip(b, (a_, b_) => a_.Equals(b_)).All(_ => _);
+    }
+}
```
Fixed in b57668b

dionjansen commented on src/DotPulsar/Internal/ConsumerChannel.cs:
```diff
@@ -108,6 +108,9 @@ public async Task Send(CommandAck command, CancellationToken cancellationToken)
 
             command.ConsumerId = _id;
+
+            _queue.Acknowledge(new MessageId(messageId));
```
This would result in that we'd need to start using `MessageIdData` for both the tracker as well as the `MessageQueue`, is this desirable? I thought the idea was we'd use `MessageId` for the tracker because they are easier to use (implements `IEquatable` and `IComparable`). If performance is more important I can refactor to `MessageIdData`

blankensteiner replied:
Ah yes, we might have to implement IEquatable and IComparable on MessageIdData also then.
When receiving an Ack, you also have to know if it is 'Individual' or 'Cumulative'.

dionjansen replied:
Ok I addressed the MessageId issue by implementing the right interfaces for comparison here 8841930, I'm not sure about the extension pattern for MessageIdData though (using partials). I refactored the use of MessageIdData in ad5827114c8d019aac86d300818b718ae8ab9d19 all tests pass.

dionjansen commented on src/DotPulsar/ConsumerOptions.cs:
```diff
@@ -101,5 +101,16 @@ public ConsumerOptions(string subscriptionName, string topic)
+        /// Delay to wait before redelivering messages that failed to be processed.
+        /// When an application uses IConsumer.NegativeAcknowledge(Message), failed messages are redelivered after a fixed timeout.
+        /// </summary>
+        public int NegativeAckRedeliveryDelayMicros { get; set; }
```
Refactored in 871dce6

dionjansen commented on src/DotPulsar/ConsumerOptions.cs:
```diff
@@ -101,5 +101,16 @@ public ConsumerOptions(string subscriptionName, string topic)
+        /// <summary>
+        /// Timeout of unacked messages
+        /// </summary>
+        public int AckTimeoutMillis { get; set; }
```
Refactored in 871dce6

dionjansen commented:
@blankensteiner thanks for the comments I managed to fix most of it with a view side comments/ questions (see above). Let me know what you think

blankensteiner commented on src/DotPulsar/Internal/InactiveUnackedMessageTracker.cs:
```diff
@@ -0,0 +1,42 @@
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public class InactiveUnackedMessageTracker : IUnackedMessageTracker
```
Let's make it sealed.

dionjansen replied:
Done: ddfd374

blankensteiner commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,158 @@
+
+    public class UnackedMessageTrackerTests
+    {
+
```
Let's remove this empty line

dionjansen replied:
Done: ddfd374

blankensteiner commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,158 @@
+            var consumer = Substitute.For<IConsumer>();
+            var messageId = MessageId.Latest;
+            var cts = new CancellationTokenSource();
+
```
Let's remove this empty line

dionjansen replied:
Done: ddfd374

blankensteiner commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,158 @@
+            var consumer = Substitute.For<IConsumer>();
+            var messageId = MessageId.Latest;
+            var cts = new CancellationTokenSource();
+
```
Let's remove this empty line

dionjansen replied:
Done: ddfd374

blankensteiner commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,158 @@
+            var consumer = Substitute.For<IConsumer>();
+            var messageId = MessageId.Latest;
+            var cts = new CancellationTokenSource();
+
```
Let's remove this empty line

dionjansen replied:
Done: ddfd374

blankensteiner commented on tests/DotPulsar.Tests/Internal/MessageAcksTrackerTests.cs:
```diff
@@ -0,0 +1,158 @@
+            var consumer = Substitute.For<IConsumer>();
+            var messageId = MessageId.Latest;
+            var cts = new CancellationTokenSource();
+
```
Let's remove this empty line

dionjansen replied:
Done: ddfd374

blankensteiner commented on src/DotPulsar/Internal/UnackedMessageTracker.cs:
```diff
@@ -0,0 +1,115 @@
+            (long) ((Stopwatch.GetTimestamp() - Timestamp) / (double) Stopwatch.Frequency * TimeSpan.TicksPerSecond));
+    }
+
+    public sealed class UnackedMessageTracker : IUnackedMessageTracker
```
For the class and interface names, let's just the full wording: [I]UnacknowledgedMessageTracner.

dionjansen replied:
Done: ddfd374

blankensteiner commented on src/DotPulsar/Internal/UnackedMessageTracker.cs:
```diff
@@ -0,0 +1,115 @@
+
+        public void Dispose()
+        {
+            this._cancellationTokenSource.Cancel();
```
No need to use "this."

dionjansen replied:
Done: ddfd374

blankensteiner commented on src/DotPulsar/Internal/Abstractions/IUnackedMessageTracker.cs:
```diff
@@ -0,0 +1,30 @@
+    {
+        void Add(MessageId messageId);
+
+        void Ack(MessageId messageId);
```
Let's use "Acknowledge" instead of the shorter "Ack", to be consistent with the public interfaces.

dionjansen replied:
Done: ddfd374

blankensteiner commented on src/DotPulsar/Internal/MessageQueue.cs:
```diff
@@ -0,0 +1,53 @@
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public sealed class MessageQueue : IMessageQueue, IDequeue<MessagePackage>, IDisposable
```
Is "IMessageQueue" enough here since it inherits the other two interfaces?

dionjansen replied:
Done: ddfd374

dionjansen commented:
@blankensteiner sorry for the delay, I addressed all your remarks, let me know what you think!

jbvanzuylen commented:
@blankensteiner @dionjansen any idea when this work will be finished and merged? Looks like a lot of work has been done and reviewed but wondering what is missing to cross the finish line.

blankensteiner commented:
Hi @jbvanzuylen 
Good question :-)
@dionjansen when you feel the PR is ready, poke me and I'll review it again :-)

dionjansen commented:
@jbvanzuylen @blankensteiner yes this dropped very far off my radar, unfortunately. I see quite a lot has changed since this implementation, I'll try to open a new PR from the latest version.

dionjansen commented:
@blankensteiner I reworked the implementation from scratch based on version [1.1.2](https://github.com/apache/pulsar-dotpulsar/releases/tag/1.1.2) and opened a new PR https://github.com/apache/pulsar-dotpulsar/pull/83. I've also added support for negative acknowledgement delays next to the unacked tracking. 

Closing this PR since it's no longer needed.

CC: @jbvanzuylen
